File: src/saturn_jolt.zig
const std = @import("std");

const options = @import("options");

const c = @cImport({
    if (options.use_double_precision) {
        @cDefine("JPH_DOUBLE_PRECISION", 1);
    }

    @cInclude("saturn_jolt.h");
});

pub fn init(allocator: std.mem.Allocator) void {
    std.debug.assert(mem_allocator == null);

    mem_allocator = allocator;
    c.init(&c.AllocationFunctions{
        .alloc = alloc,
        .free = free,
        .aligned_alloc = alignedAlloc,
        .aligned_free = free,
        .realloc = reallocate,
    });
}
pub fn deinit() void {
    c.deinit();
    mem_allocator = null;
}

// Types
pub const RVec3 = c.RVec3;
pub const Vec3 = c.Vec3;
pub const Quat = c.Quat;

pub const Transform = c.Transform;
pub const Velocity = c.Velocity;

pub const UserData = c.UserData;
pub const ObjectLayer = c.ObjectLayer;
pub const SubShapeIndex = c.SubShapeIndex;
pub const MassProperties = c.MassProperties;

pub const RayCastHit = c.RayCastHit;
pub const ShapeCastHit = c.ShapeCastHit;
pub const ShapeCastCallback = c.ShapeCastCallback;

pub const SubShapeSettings = struct {
    shape: Shape,
    position: Vec3,
    rotation: Quat,
};

pub const Shape = struct {
    const Self = @This();

    handle: c.Shape,

    pub fn initSphere(radius: f32, density: f32, user_data: UserData) Self {
        return .{
            .handle = c.shapeCreateSphere(radius, density, user_data),
        };
    }

    pub fn initBox(half_extent: [3]f32, density: f32, user_data: UserData) Self {
        return .{
            .handle = c.shapeCreateBox(&half_extent, density, user_data),
        };
    }

    pub fn initCylinder(half_height: f32, radius: f32, density: f32, user_data: UserData) Self {
        return .{
            .handle = c.shapeCreateCylinder(half_height, radius, density, user_data),
        };
    }

    pub fn initCapsule(half_height: f32, radius: f32, density: f32, user_data: UserData) Self {
        return .{
            .handle = c.shapeCreateCapsule(half_height, radius, density, user_data),
        };
    }

    pub fn initConvexHull(positions: [][3]f32, density: f32, user_data: UserData) Self {
        return .{
            .handle = c.shapeCreateConvexHull(@alignCast(@ptrCast(positions.ptr)), positions.len, density, user_data),
        };
    }

    pub fn initMeshStatic(positions: [][3]f32, indices: []u32, user_data: UserData) Self {
        return .{
            .handle = c.shapeCreateMesh(@alignCast(@ptrCast(positions.ptr)), positions.len, @alignCast(@ptrCast(indices.ptr)), indices.len, null, user_data),
        };
    }

    pub fn initMeshWithMass(positions: [][3]f32, indices: []u32, mass_properites: MassProperties, user_data: UserData) Self {
        return .{
            .handle = c.shapeCreateMesh(@alignCast(@ptrCast(positions.ptr)), positions.len, @alignCast(@ptrCast(indices.ptr)), indices.len, &mass_properites, user_data),
        };
    }

    pub fn initCompound(sub_shapes: []const SubShapeSettings, user_data: UserData) Self {
        //TODO: since Shape is currently a wrapper type, we can cast it to a c.SubShapeSettings without issue, may become a problem on diffrent platforms
        comptime std.debug.assert(@sizeOf(SubShapeSettings) == @sizeOf(c.SubShapeSettings));
        return .{
            .handle = c.shapeCreateCompound(@alignCast(@ptrCast(sub_shapes.ptr)), sub_shapes.len, user_data),
        };
    }

    pub fn deinit(self: *Self) void {
        c.shapeDestroy(self.handle);
    }

    pub fn getMassProperties(self: Self) MassProperties {
        return c.shapeGetMassProperties(self.handle);
    }
};

pub const MotionType = enum(u32) {
    static = 0,
    kinematic = 1,
    dynamic = 2,
};

pub const World = struct {
    const Self = @This();

    pub const Settings = struct {
        max_bodies: u32 = 1024,
        num_body_mutexes: u32 = 0,
        max_body_pairs: u32 = 1024,
        max_contact_constraints: u32 = 1024,
        temp_allocation_size: u32 = 1024 * 1024 * 10,
        threads: u16 = 4,
    };

    ptr: ?*c.World,

    pub fn init(settings: Settings) Self {
        return .{
            .ptr = c.worldCreate(&.{
                .max_bodies = settings.max_bodies,
                .num_body_mutexes = settings.num_body_mutexes,
                .max_body_pairs = settings.max_body_pairs,
                .max_contact_constraints = settings.max_contact_constraints,
                .temp_allocation_size = settings.temp_allocation_size,
                .threads = settings.threads,
            }),
        };
    }

    pub fn deinit(self: *Self) void {
        c.worldDestroy(self.ptr);
    }

    pub fn addBody(self: *Self, body: Body) void {
        c.worldAddBody(self.ptr, body.ptr);
    }

    pub fn removeBody(self: *Self, body: Body) void {
        c.worldRemoveBody(self.ptr, body.ptr);
    }

    pub fn update(self: *Self, delta_time: f32, collisions_steps: i32) void {
        c.worldUpdate(self.ptr, delta_time, collisions_steps);
    }

    pub fn castRayClosest(self: Self, object_layer_pattern: ObjectLayer, origin: RVec3, direction: Vec3) ?RayCastHit {
        var hit: c.RayCastHit = .{};
        return if (c.worldCastRayAll(self.ptr, object_layer_pattern, @ptrCast(&origin[0]), @ptrCast(&direction[0]), &hit))
            hit
        else
            null;
    }

    pub fn castRayClosestIgnoreBody(self: Self, object_layer_pattern: ObjectLayer, ignore_body: Body, origin: RVec3, direction: Vec3) ?c.RayCastHit {
        var hit: c.RayCastHit = .{};
        return if (c.worldCastRayClosetIgnoreBody(self.ptr, object_layer_pattern, ignore_body.ptr, @ptrCast(&origin[0]), @ptrCast(&direction[0]), &hit))
            hit
        else
            null;
    }

    pub fn castShape(self: Self, object_layer_pattern: ObjectLayer, shape: Shape, transform: *const Transform, callback: ShapeCastCallback, callback_data: *anyopaque) void {
        c.worldCastShape(self.ptr, object_layer_pattern, shape.handle, transform, callback, callback_data);
    }
};

pub const Body = struct {
    const Self = @This();

    pub const Settings = struct {
        transform: Transform = .{
            .position = .{ 0.0, 0.0, 0.0 },
            .rotation = .{ 0.0, 0.0, 0.0, 1.0 },
        },
        linear_velocity: Vec3 = .{ 0.0, 0.0, 0.0 },
        angular_velocity: Vec3 = .{ 0.0, 0.0, 0.0 },
        user_data: UserData = 0,
        object_layer: ObjectLayer,
        motion_type: MotionType,
        allow_sleep: bool = true,
        friction: f32 = 0.0,
        linear_damping: f32 = 0.0,
        angular_damping: f32 = 0.0,
        gravity_factor: f32 = 1.0,
        max_linear_velocity: f32 = 500.0,
        max_angular_velocity: f32 = std.math.pi * 4.0,
    };

    ptr: ?*c.Body,

    pub fn init(settings: Settings) Self {
        return .{
            .ptr = c.bodyCreate(&.{
                .position = settings.transform.position,
                .rotation = settings.transform.rotation,
                .linear_velocity = settings.linear_velocity,
                .angular_velocity = settings.angular_velocity,
                .user_data = settings.user_data,
                .object_layer = settings.object_layer,
                .motion_type = @intFromEnum(settings.motion_type),
                .allow_sleep = settings.allow_sleep,
                .friction = settings.friction,
                .linear_damping = settings.angular_damping,
                .gravity_factor = settings.gravity_factor,
                .max_linear_velocity = settings.max_linear_velocity,
                .max_angular_velocity = settings.max_angular_velocity,
            }),
        };
    }

    pub fn deinit(self: *Self) void {
        c.bodyDestroy(self.ptr);
    }

    pub fn getWorld(self: *Self) ?World {
        if (c.bodyGetWorld(self.ptr)) |world_ptr| {
            return .{ .ptr = world_ptr };
        }
        return null;
    }

    pub fn getTransform(self: *Self) Transform {
        return c.bodyGetTransform(self.ptr);
    }

    pub fn setTransform(self: *Self, transform: *const Transform) void {
        c.bodySetTransform(self.ptr, transform);
    }

    pub fn getVelocity(self: *Self) Velocity {
        return c.bodyGetVelocity(self.ptr);
    }

    pub fn setVelocity(self: *Self, velocity: *const Velocity) void {
        c.bodySetVelocity(self.ptr, velocity);
    }

    pub fn addForce(self: *Self, force: Vec3, activate: bool) void {
        c.bodyAddForce(self.ptr, @ptrCast(&force[0]), activate);
    }

    pub fn addTorque(self: *Self, torque: Vec3, activate: bool) void {
        c.bodyAddTorque(self.ptr, @ptrCast(&torque[0]), activate);
    }

    pub fn addImpulse(self: *Self, impulse: Vec3) void {
        c.bodyAddImpulse(self.ptr, @ptrCast(&impulse[0]));
    }

    pub fn addAngularImpulse(self: *Self, angular_impulse: Vec3) void {
        c.bodyAddAngularImpulse(self.ptr, @ptrCast(&angular_impulse[0]));
    }

    pub fn addShape(self: *Self, shape: Shape, position: Vec3, rotation: Quat, user_data: UserData) SubShapeIndex {
        return c.bodyAddShape(self.ptr, shape.handle, @ptrCast(&position[0]), @ptrCast(&rotation[0]), user_data);
    }

    pub fn removeShape(self: *Self, index: SubShapeIndex) void {
        return c.bodyRemoveShape(self.ptr, index);
    }

    pub fn updateShapeTransform(self: *Self, index: SubShapeIndex, position: Vec3, rotation: Quat) void {
        return c.bodyUpdateShapeTransform(self.ptr, index, @ptrCast(&position[0]), @ptrCast(&rotation[0]));
    }

    pub fn removeAllShapes(self: *Self) void {
        return c.bodyRemoveAllShapes(self.ptr);
    }

    pub fn commitShapeChanges(self: *Self) void {
        return c.bodyCommitShapeChanges(self.ptr);
    }
};

pub const MeshPrimitive = c.MeshPrimitive;
pub const Vertex = c.Vertex;
pub const Triangle = c.Triangle;
pub const DebugRendererCallbacks = c.DebugRendererCallbacks;
pub const DrawLineData = c.DrawLineData;
pub const DrawTriangleData = c.DrawTriangleData;
pub const DrawTextData = c.DrawTextData;
pub const DrawGeometryData = c.DrawGeometryData;

pub const CullMode = enum(c.CullMode) {
    back_face = c.CULL_MODE_BACK,
    front_face = c.CULL_MODE_FRONT,
    off = c.CULL_MODE_OFF,
};

pub const DrawMode = enum(c.DrawMode) {
    solid = c.DRAW_MODE_SOLID,
    wireframe = c.DRAW_MODE_WIREFRAME,
};

pub fn initDebugRenderer(debug_renderer: c.DebugRendererCallbacks) void {
    c.debugRendererCreate(debug_renderer);
}

pub fn deinitDebugRenderer() void {
    c.debugRendererDestroy();
}

pub fn debugRendererBuildFrame(world: *World, transform: Transform, ignore_bodies: []const Body) void {
    c.debugRendererBuildFrame(world.ptr, transform, @ptrCast(ignore_bodies.ptr), @intCast(ignore_bodies.len));
}

// Memory Allocation
var mem_allocator: ?std.mem.Allocator = null;
const Metadata = struct {
    size: usize,
    alignment: usize,
};

fn alloc(size: usize) callconv(.C) ?*anyopaque {
    return alignedAlloc(size, 16);
}

fn alignedAlloc(size: usize, alignment: usize) callconv(.C) ?*anyopaque {
    if (mem_allocator) |allocator| {
        const allocation_size = size + alignment;

        const allocation: []u8 = switch (alignment) {
            16 => allocator.alignedAlloc(u8, 16, allocation_size),
            32 => allocator.alignedAlloc(u8, 32, allocation_size),
            64 => allocator.alignedAlloc(u8, 64, allocation_size),
            else => |x| std.debug.panic("Unsupported memory aligment: {}", .{x}),
        } catch return null;

        const data_ptr_int: usize = @as(usize, @intFromPtr(allocation.ptr)) + alignment;
        const data_ptr: *u8 = @ptrFromInt(data_ptr_int);
        const metadata_ptr: *Metadata = @ptrFromInt(data_ptr_int - @sizeOf(Metadata));
        metadata_ptr.size = size;
        metadata_ptr.alignment = alignment;

        return data_ptr;
    }
    return null;
}

fn reallocate(maybe_ptr: ?*anyopaque, old_size: usize, new_size: usize) callconv(.C) ?*anyopaque {
    if (maybe_ptr) |data_ptr| {
        if (mem_allocator) |allocator| {
            const data_ptr_int: usize = @intFromPtr(data_ptr);
            const metadata_ptr: *Metadata = @ptrFromInt(data_ptr_int - @sizeOf(Metadata));
            const allocation_ptr: [*]u8 = @ptrFromInt(data_ptr_int - metadata_ptr.alignment);
            const allocation_size: usize = metadata_ptr.size + metadata_ptr.alignment;

            if (old_size != metadata_ptr.size) {
                std.log.warn("Jolt expected memory size({}) doesn't match memory's metadata({}), there may be a bug in the allocator", .{ old_size, metadata_ptr.size });
            }

            const new_allocation_size = new_size + metadata_ptr.alignment;

            // Try to resize current buffer
            const resized: bool = switch (metadata_ptr.alignment) {
                16 => allocator.resize(@as([*]align(16) u8, @alignCast(allocation_ptr))[0..allocation_size], new_allocation_size),
                32 => allocator.resize(@as([*]align(32) u8, @alignCast(allocation_ptr))[0..allocation_size], new_allocation_size),
                64 => allocator.resize(@as([*]align(64) u8, @alignCast(allocation_ptr))[0..allocation_size], new_allocation_size),
                else => |x| std.debug.panic("Unsupported memory aligment: {}", .{x}),
            };
            if (resized) {
                metadata_ptr.size = new_size;
                return data_ptr;
            }

            // If resize fails, try allocate and memcopy new buffer
            if (alignedAlloc(new_size, metadata_ptr.alignment)) |new_ptr| {
                defer free(maybe_ptr);

                if (new_size != 0) {
                    var old_data_ptr_bytes: [*]u8 = @ptrCast(data_ptr);
                    var new_data_ptr_bytes: [*]u8 = @ptrCast(new_ptr);
                    const copy_size = @min(metadata_ptr.size, new_size);
                    @memcpy(new_data_ptr_bytes[0..copy_size], old_data_ptr_bytes[0..copy_size]);
                    return new_ptr;
                } else {
                    return null;
                }
            }
        }
    }
    return alloc(new_size);
}

fn free(maybe_ptr: ?*anyopaque) callconv(.C) void {
    if (maybe_ptr) |data_ptr| {
        if (mem_allocator) |allocator| {
            const data_ptr_int: usize = @intFromPtr(data_ptr);
            const metadata_ptr: *Metadata = @ptrFromInt(data_ptr_int - @sizeOf(Metadata));
            const allocation_ptr: [*]u8 = @ptrFromInt(data_ptr_int - metadata_ptr.alignment);
            const allocation_size: usize = metadata_ptr.size + metadata_ptr.alignment;

            switch (metadata_ptr.alignment) {
                16 => allocator.free(@as([*]align(16) u8, @alignCast(allocation_ptr))[0..allocation_size]),
                32 => allocator.free(@as([*]align(32) u8, @alignCast(allocation_ptr))[0..allocation_size]),
                64 => allocator.free(@as([*]align(64) u8, @alignCast(allocation_ptr))[0..allocation_size]),
                else => |x| std.debug.panic("Unsupported memory aligment: {}", .{x}),
            }
        }
    }
}

File: src/saturn_jolt.cpp
#include "saturn_jolt.h"

#include <cstdio>
#include <iostream>
#include <thread>

#include <Jolt/Jolt.h>

#include <Jolt/Core/Factory.h>
#include <Jolt/Physics/Collision/CastResult.h>
#include <Jolt/Physics/Collision/CollideShape.h>
#include <Jolt/Physics/Collision/RayCast.h>
#include <Jolt/Physics/PhysicsSystem.h>
#include <Jolt/RegisterTypes.h>

#include "memory.hpp"
#include "shape_pool.hpp"

#include "Jolt/Physics/Collision/Shape/BoxShape.h"
#include "Jolt/Physics/Collision/Shape/CapsuleShape.h"
#include "Jolt/Physics/Collision/Shape/ConvexHullShape.h"
#include "Jolt/Physics/Collision/Shape/CylinderShape.h"
#include "Jolt/Physics/Collision/Shape/MeshShape.h"
#include "Jolt/Physics/Collision/Shape/SphereShape.h"
#include "Jolt/Physics/Collision/Shape/StaticCompoundShape.h"
#include "body.hpp"
#include "mass_shape.hpp"
#include "world.hpp"

ShapePool *shape_pool = nullptr;
std::mutex shape_pool_mutex;

void init(const AllocationFunctions *functions) {
	if (functions != nullptr) {
		JPH::Allocate = functions->alloc;
		JPH::Free = functions->free;
		JPH::AlignedAllocate = functions->aligned_alloc;
		JPH::AlignedFree = functions->aligned_free;
		JPH::Reallocate = functions->realloc;
	} else {
		JPH::RegisterDefaultAllocator();
	}

	auto factory =
		static_cast<JPH::Factory *>(JPH::Allocate(sizeof(JPH::Factory)));
	::new (factory) JPH::Factory();
	JPH::Factory::sInstance = factory;
	JPH::RegisterTypes();

	JPH::MassShape::sRegister();

	shape_pool = new ShapePool();
}

void deinit() {
	debugRendererDestroy();
	shape_pool_mutex.lock();
	delete shape_pool;
	shape_pool_mutex.unlock();

	JPH::UnregisterTypes();
	JPH::Factory::sInstance->~Factory();
	JPH::Free((void *)JPH::Factory::sInstance);
	JPH::Factory::sInstance = nullptr;

	JPH::Allocate = nullptr;
	JPH::Free = nullptr;
	JPH::AlignedAllocate = nullptr;
	JPH::AlignedFree = nullptr;
}

#ifdef JPH_DEBUG_RENDERER
#include "callback_debug_renderer.hpp"

Color color_from_u32(uint32_t value) {
	Color color;
	color.r = (value >> 24) & 0xFF;
	color.g = (value >> 16) & 0xFF;
	color.b = (value >> 8) & 0xFF;
	color.a = value & 0xFF;
	return color;
}
#endif

void debugRendererCreate(DebugRendererCallbacks data) {
#ifdef JPH_DEBUG_RENDERER
	new CallbackDebugRenderer(data);
#endif
}
void debugRendererDestroy() {
#ifdef JPH_DEBUG_RENDERER
	delete JPH::DebugRenderer::sInstance;
#endif
}

void debugRendererBuildFrame(World *world_ptr, Transform camera_transform, const Body* const* ignore_bodies, const uint32_t ignore_bodies_count) {
#ifdef JPH_DEBUG_RENDERER
	if (JPH::DebugRenderer::sInstance != nullptr && world_ptr != nullptr) {
		auto renderer = dynamic_cast<CallbackDebugRenderer *>(JPH::DebugRenderer::sInstance);
		renderer->camera_position = loadRVec3(camera_transform.position);

		IgnoreListBodyDrawFilter draw_filter(ignore_bodies, ignore_bodies_count);

		JPH::BodyManager::DrawSettings settings;
		settings.mDrawShape = true;
		settings.mDrawShapeWireframe = true;
		settings.mDrawShapeColor = JPH::BodyManager::EShapeColor::SleepColor;
		world_ptr->physics_system->DrawBodies(settings, JPH::DebugRenderer::sInstance, &draw_filter);
	}
#endif
}

Shape shapeCreateSphere(float radius, float density, UserData user_data) {
	auto settings = JPH::SphereShapeSettings();
	settings.mRadius = radius;
	settings.mDensity = density;
	settings.mUserData = user_data;
	auto shape = settings.Create().Get();
	shape_pool_mutex.lock();
	auto shape_handle = shape_pool->insert(shape);
	shape_pool_mutex.unlock();
	return shape_handle;
}

Shape shapeCreateBox(const Vec3 half_extent, float density, UserData user_data) {
	auto settings = JPH::BoxShapeSettings();
	settings.mHalfExtent = loadVec3(half_extent);
	settings.mDensity = density;
	settings.mUserData = user_data;

	auto shape = settings.Create().Get();
	shape_pool_mutex.lock();
	auto shape_handle = shape_pool->insert(shape);
	shape_pool_mutex.unlock();
	return shape_handle;
}

Shape shapeCreateCylinder(float half_height, float radius, float density, UserData user_data) {
	auto settings = JPH::CylinderShapeSettings();
	settings.mHalfHeight = half_height;
	settings.mRadius = radius;
	settings.mDensity = density;
	settings.mUserData = user_data;

	auto shape = settings.Create().Get();
	shape_pool_mutex.lock();
	auto shape_handle = shape_pool->insert(shape);
	shape_pool_mutex.unlock();
	return shape_handle;
}

Shape shapeCreateCapsule(float half_height, float radius, float density, UserData user_data) {
	auto settings = JPH::CapsuleShapeSettings();
	settings.mHalfHeightOfCylinder = half_height;
	settings.mRadius = radius;
	settings.mDensity = density;
	settings.mUserData = user_data;

	auto shape = settings.Create().Get();
	shape_pool_mutex.lock();
	auto shape_handle = shape_pool->insert(shape);
	shape_pool_mutex.unlock();
	return shape_handle;
}

Shape shapeCreateConvexHull(const Vec3 positions[], size_t position_count, float density, UserData user_data) {
	JPH::Array<JPH::Vec3> point_list(position_count);
	for (size_t i = 0; i < position_count; i++) {
		point_list[i] = loadVec3(positions[i]);
	}
	auto settings = JPH::ConvexHullShapeSettings();
	settings.mPoints = point_list;
	settings.mDensity = density;
	settings.mUserData = user_data;

	auto shape = settings.Create().Get();
	shape_pool_mutex.lock();
	auto shape_handle = shape_pool->insert(shape);
	shape_pool_mutex.unlock();
	return shape_handle;
}

Shape shapeCreateMesh(const Vec3 positions[], size_t position_count, const uint32_t *indices, size_t indices_count, const MassProperties *mass_properties, UserData user_data) {
	JPH::MassProperties mass;

	JPH::VertexList vertex_list;
	for (size_t i = 0; i < position_count; i++) {
		vertex_list.push_back(loadFloat3(positions[i]));
	}

	JPH::IndexedTriangleList triangle_list;

	if (indices_count == 0) {
		for (int i = 0; i < position_count; i += 3) {
			triangle_list.push_back(JPH::IndexedTriangle(i + 0, i + 1, i + 2, 0));
		}
	} else {
		const size_t triangle_count = indices_count / 3;
		for (int i = 0; i < triangle_count; i++) {
			const size_t offset = i * 3;
			triangle_list.push_back(JPH::IndexedTriangle(
				indices[offset + 0], indices[offset + 1], indices[offset + 2], 0));
		}
	}

	auto settings = JPH::MeshShapeSettings(vertex_list, triangle_list);
	settings.mUserData = user_data;

	auto shape = settings.Create().Get();

	if (mass_properties != nullptr) {
		JPH::MassProperties override_mass;
		override_mass.mMass = mass_properties->mass;
		override_mass.mInertia = JPH::Mat44::sLoadFloat4x4(reinterpret_cast<const JPH::Float4 *>(mass_properties->inertia_tensor));
		shape = JPH::MassShapeSettings(shape, override_mass).Create().Get();
	}

	shape_pool_mutex.lock();
	auto shape_handle = shape_pool->insert(shape);
	shape_pool_mutex.unlock();
	return shape_handle;
}

Shape shapeCreateCompound(const SubShapeSettings sub_shapes[], size_t sub_shape_count, UserData user_data) {
	auto static_shape_settings = JPH::StaticCompoundShapeSettings();
	static_shape_settings.mUserData = user_data;

	shape_pool_mutex.lock();
	for (auto i = 0; i < sub_shape_count; i++) {
		auto subshape = &sub_shapes[i];
		auto shape_ref = shape_pool->get(subshape->shape);
		static_shape_settings.AddShape(loadVec3(subshape->position), loadQuat(subshape->rotation), shape_ref, i);
	}

	auto shape = static_shape_settings.Create().Get();
	auto shape_handle = shape_pool->insert(shape);
	shape_pool_mutex.unlock();
	return shape_handle;
}

void shapeDestroy(Shape shape) {
	shape_pool_mutex.lock();
	shape_pool->remove(shape);
	shape_pool_mutex.unlock();
}

MassProperties shapeGetMassProperties(Shape shape) {
	shape_pool_mutex.lock();
	auto shape_ref = shape_pool->get(shape);
	shape_pool_mutex.unlock();
	auto shape_properties = shape_ref->GetMassProperties();
	MassProperties mass_properties;
	mass_properties.mass = shape_properties.mMass;
	shape_properties.mInertia.StoreFloat4x4(reinterpret_cast<JPH::Float4 *>(mass_properties.inertia_tensor));
	return mass_properties;
}

World *worldCreate(const WorldSettings *settings) {
	return new World(settings);
}

void worldDestroy(World *world_ptr) {
	delete world_ptr;
}

void worldUpdate(World *world_ptr, float delta_time, int collision_steps) {
	world_ptr->update(delta_time, collision_steps);
}

void worldAddBody(World *world_ptr, Body *body_ptr) {
	world_ptr->addBody(body_ptr);
}

void worldRemoveBody(World *world_ptr, Body *body_ptr) {
	world_ptr->removeBody(body_ptr);
}

bool worldCastRayCloset(World *world_ptr, ObjectLayer object_layer_pattern, const RVec3 origin, const Vec3 direction, RayCastHit *hit_result) {
	return world_ptr->castRayCloset(object_layer_pattern, loadRVec3(origin), loadVec3(direction), hit_result);
}

bool worldCastRayClosetIgnoreBody(World *world_ptr, ObjectLayer object_layer_pattern, Body *ignore_body_ptr, const RVec3 origin, const Vec3 direction, RayCastHit *hit_result) {
	JPH::BodyID ignore_body_id;

	if (ignore_body_ptr != nullptr && ignore_body_ptr->world_ptr == world_ptr) {
		ignore_body_id = ignore_body_ptr->body_id;
	}

	return world_ptr->castRayClosetIgnoreBody(object_layer_pattern, ignore_body_id, loadRVec3(origin), loadVec3(direction), hit_result);
}

void worldCastShape(World *world_ptr, ObjectLayer object_layer_pattern, Shape shape, const Transform *c_transform, ShapeCastCallback callback, void *callback_data) {
	shape_pool_mutex.lock();
	auto shape_ref = shape_pool->get(shape);
	shape_pool_mutex.unlock();
	world_ptr->castShape(object_layer_pattern, loadRVec3(c_transform->position), loadQuat(c_transform->rotation), shape_ref, callback, callback_data);
}

// Body functions
Body *bodyCreate(const BodySettings *settings) {
	return new Body(settings);
}

void bodyDestroy(Body *body_ptr) {
	delete body_ptr;
}

World *bodyGetWorld(Body *body_ptr) {
	return body_ptr->getWorld();
}

Transform bodyGetTransform(Body *body_ptr) {
	auto position = body_ptr->getPosition();
	auto rotation = body_ptr->getRotation();
	return Transform{
		{position.GetX(), position.GetY(), position.GetZ()},
		{rotation.GetX(), rotation.GetY(), rotation.GetZ(), rotation.GetW()}};
}

void bodySetTransform(Body *body_ptr, const Transform *c_transform) {
	body_ptr->setTransform(loadRVec3(c_transform->position), loadQuat(c_transform->rotation));
}

Velocity bodyGetVelocity(Body *body_ptr) {
	auto linear = body_ptr->getLinearVelocity();
	auto angular = body_ptr->getAngularVelocity();
	return Velocity{
		{linear.GetX(), linear.GetY(), linear.GetZ()},
		{angular.GetX(), angular.GetY(), angular.GetZ()},

	};
}

void bodySetVelocity(Body *body_ptr, const Velocity *c_velocity) {
	body_ptr->setVelocity(loadVec3(c_velocity->linear), loadVec3(c_velocity->angular));
}

void bodyAddForce(Body *body_ptr, const Vec3 force, bool activate) {
	body_ptr->addForce(loadVec3(force), activate ? JPH::EActivation::Activate : JPH::EActivation::DontActivate);
}

void bodyAddTorque(Body *body_ptr, const Vec3 torque, bool activate) {
	body_ptr->addTorque(loadVec3(torque), activate ? JPH::EActivation::Activate : JPH::EActivation::DontActivate);
}

void bodyAddImpulse(Body *body_ptr, const Vec3 impulse) {
	body_ptr->addImpulse(loadVec3(impulse));
}

void bodyAddAngularImpulse(Body *body_ptr, const Vec3 angular_impulse) {
	body_ptr->addAngularImpulse(loadVec3(angular_impulse));
}

SubShapeIndex bodyAddShape(Body *body_ptr, Shape shape, const Vec3 position, const Quat rotation, UserData user_data) {
	shape_pool_mutex.lock();
	auto shape_ref = shape_pool->get(shape);
	shape_pool_mutex.unlock();

	return body_ptr->addShape(SubShape{
		shape_ref,
		loadVec3(position),
		loadQuat(rotation),
		user_data,
	});
}

void bodyRemoveShape(Body *body_ptr, SubShapeIndex index) {
	body_ptr->removeShape(index);
}

void bodyUpdateShapeTransform(Body *body_ptr, SubShapeIndex index, const Vec3 position, const Quat rotation) {
	body_ptr->updateShapeTransform(index, loadVec3(position), loadQuat(rotation));
}

void bodyRemoveAllShapes(Body *body_ptr) {
	body_ptr->removeAllShape();
}

void bodyCommitShapeChanges(Body *body_ptr) {
	body_ptr->commitShapeChanges();
}
